File: app/src/main/java/com/alexkn/syntact/data/model/views/BucketDetail.java
package com.alexkn.syntact.data.model.views;

import androidx.room.DatabaseView;

import com.alexkn.syntact.data.common.Identifiable;

import java.time.Instant;
import java.util.Locale;
import java.util.Objects;

@DatabaseView(
        "SELECT b.id, b.language, b.createdAt,b.itemCount, (SELECT count(*) FROM solvableitem s JOIN bucket b  ON s.bucketId = b.id WHERE s.timesSolved > 0) as solvedCount FROM Bucket b;")
public class BucketDetail implements Identifiable {

    private Long id;

    private Locale language;

    private Instant createdAt;

    private Integer itemCount;

    private Integer solvedCount;

    public Long getId() {

        return id;
    }

    public void setId(Long id) {

        this.id = id;
    }

    public Locale getLanguage() {

        return language;
    }

    public void setLanguage(Locale language) {

        this.language = language;
    }

    public Instant getCreatedAt() {

        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {

        this.createdAt = createdAt;
    }

    public Integer getItemCount() {

        return itemCount;
    }

    public void setItemCount(Integer itemCount) {

        this.itemCount = itemCount;
    }

    public Integer getSolvedCount() {

        return solvedCount;
    }

    public void setSolvedCount(Integer solvedCount) {

        this.solvedCount = solvedCount;
    }

    @Override
    public boolean equals(Object o) {

        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        BucketDetail that = (BucketDetail) o;
        return Objects.equals(id, that.id) && Objects.equals(language, that.language) && Objects.equals(createdAt, that.createdAt) &&
                Objects.equals(itemCount, that.itemCount) && Objects.equals(solvedCount, that.solvedCount);
    }

    @Override
    public int hashCode() {

        return Objects.hash(id, language, createdAt, itemCount, solvedCount);
    }
}

File: app/src/main/java/com/alexkn/syntact/data/model/views/BucketDetail.kt
package com.alexkn.syntact.data.model.views

import androidx.room.DatabaseView
import com.alexkn.syntact.data.common.Identifiable
import java.time.Instant
import java.util.*

@DatabaseView("SELECT b.name, b.id, b.language, b.createdAt,b.itemCount, (SELECT count(*) FROM solvableitem s WHERE s.timesSolved > 0 AND s.disabled = 0 AND s.bucketId = b.id) as solvedCount, (SELECT count(*) FROM solvableitem s WHERE s.bucketId = b.id AND s.disabled = 0) as onDeviceCount,(SELECT b.itemCount - (SELECT count(*) FROM solvableitem s WHERE s.nextDueDate > (SELECT strftime('%s','now') || substr(strftime('%f','now'),4)) AND s.bucketId = b.id) - (SELECT count(*) FROM solvableitem s WHERE s.disabled = 1 AND s.nextDueDate <= (SELECT strftime('%s','now') || substr(strftime('%f','now'),4)) AND s.bucketId = b.id)) as dueCount,  (SELECT count(*) FROM solvableitem s WHERE s.disabled = 1 AND s.bucketId = b.id) as disabledCount FROM Bucket b;")
data class BucketDetail(
        override var id: Long,
        var name: String,
        var language: Locale,
        var createdAt: Instant,
        var itemCount: Int,
        var solvedCount: Int,
        var onDeviceCount: Int,
        var dueCount: Int,
        var disabledCount: Int
) : Identifiable<Long>
